File: samlang-parser-generated-ts/src/statement-block-builder.ts
import { PLVisitor } from './generated/PLVisitor';
import { AbstractParseTreeVisitor } from 'antlr4ts/tree/AbstractParseTreeVisitor';
import {
  ExpressionContext,
  StatementBlockContext,
  ValStatementContext,
} from './generated/PLParser';
import { UndecidedType, Expression, StatementBlock, ValStatement, WildCardPattern } from './ast';
import { contextRange, throwParserError } from './parser-util';
import typeBuilder from './type-builder';
import patternBuilder from './pattern-builder';

class StatementBuilder extends AbstractParseTreeVisitor<ValStatement>
  implements PLVisitor<ValStatement> {
  constructor(private readonly expressionBuilder: (context: ExpressionContext) => Expression) {
    super();
  }

  defaultResult = (): ValStatement => throwParserError();

  visitValStatement = (ctx: ValStatementContext): ValStatement => {
    const patternContext = ctx.pattern() ?? throwParserError();
    const pattern =
      patternContext.accept(patternBuilder) ?? new WildCardPattern(contextRange(patternContext));
    const typeAnnotation =
      ctx.typeAnnotation()?.typeExpr()?.accept(typeBuilder) ?? new UndecidedType();
    const expressionContext = ctx.expression() ?? throwParserError();
    const expression = this.expressionBuilder(expressionContext) ?? throwParserError();
    return new ValStatement(contextRange(ctx), pattern, typeAnnotation, expression);
  };
}

export default class StatementBlockBuilder extends AbstractParseTreeVisitor<StatementBlock>
  implements PLVisitor<StatementBlock> {
  private readonly statementBuilder: StatementBuilder;

  constructor(private readonly expressionBuilder: (context: ExpressionContext) => Expression) {
    super();
    this.statementBuilder = new StatementBuilder(expressionBuilder);
  }

  defaultResult = (): StatementBlock => throwParserError();

  visitStatementBlock = (ctx: StatementBlockContext): StatementBlock => {
    const expressionContext = ctx.expression() ?? throwParserError();
    const expression = this.expressionBuilder(expressionContext) ?? throwParserError();
    return new StatementBlock(
      contextRange(ctx),
      ctx.statement().map((it) => it.accept(this.statementBuilder)),
      expression
    );
  };
}

File: samlang-parser/src/jvmMain/kotlin/samlang/parser/StatementBlockBuilder.kt
package samlang.parser

import samlang.ast.common.Type
import samlang.ast.lang.Expression
import samlang.ast.lang.Pattern
import samlang.ast.lang.Statement
import samlang.ast.lang.StatementBlock
import samlang.parser.generated.PLBaseVisitor
import samlang.parser.generated.PLParser

internal class StatementBlockBuilder(
    private val expressionBuilder: (PLParser.ExpressionContext) -> Expression
) : PLBaseVisitor<StatementBlock?>() {
    private val statementBuilder: StatementBuilder = StatementBuilder()

    override fun visitStatementBlock(ctx: PLParser.StatementBlockContext): StatementBlock? =
        StatementBlock(
            range = ctx.range,
            statements = ctx.statement().mapNotNull { it.accept(statementBuilder) },
            expression = ctx.expression()?.let(block = expressionBuilder)
        )

    private inner class StatementBuilder : PLBaseVisitor<Statement?>() {

        override fun visitValStatement(ctx: PLParser.ValStatementContext): Statement? {
            val pattern = ctx.pattern()?.let { patternContext ->
                patternContext.accept(PatternBuilder) ?: Pattern.WildCardPattern(range = patternContext.range)
            } ?: return null
            val typeAnnotation = ctx.typeAnnotation()?.typeExpr()?.accept(TypeBuilder) ?: Type.undecided()
            val expression = ctx.expression()?.let(block = expressionBuilder) ?: return null
            return Statement.Val(
                range = ctx.range,
                pattern = pattern,
                typeAnnotation = typeAnnotation,
                assignedExpression = expression
            )
        }
    }
}
